core/src/sumeragi/network_topology.rs

```diff
@@ -5,341 +5,183 @@
-        self.peers = Some(peers.into_iter().collect());
-        self
-    }
-
-    /// Set the latest committed block.
-    pub(crate) fn at_block(mut self, block: HashOf<VersionedCommittedBlock>) -> Self {
-        self.at_block = Some(block);
-        self
-    }
-
-    /// Build and get topology.
-    ///
-    /// # Errors
-    /// 1. Required field is omitted.
-    /// 2. No peer exists.
-    pub(crate) fn build(self, number_of_view_changes_this_round: u64) -> Result<Topology> {
-        let peers = field_is_some_or_err!(self.peers)?;
-        if peers.is_empty() {
-            return Err(eyre!("There must be at least one peer in the network."));
-        }
-        let at_block = self.at_block;
-        let peers: Vec<_> = peers.into_iter().collect();
-        let n_view_changes = number_of_view_changes_this_round;
-        let since_last_shuffle = n_view_changes % (peers.len() as u64);
-        let is_full_circle = since_last_shuffle == 0;
-        let sorted_peers = if is_full_circle {
-            sort_peers_by_hash_and_counter(peers, &at_block, n_view_changes)
-        } else {
-            let last_shuffled_at = n_view_changes - since_last_shuffle;
-            let peers = sort_peers_by_hash_and_counter(peers, &at_block, last_shuffled_at);
-            shift_peers_by_n(peers, since_last_shuffle)
-        };
-        Ok(Topology { sorted_peers })
-    }
-}
-
 /// The ordering of the peers which defines their roles in the current round of consensus.
-#[derive(Clone, Debug, Encode, Decode, IntoSchema, Serialize)]
+///
+/// A  |       |              |>|                  |->|
+/// B  |       |              | |                  |  V
+/// C  | A Set |              ^ V  Rotate A Set    ^  |
+/// D  | 2f +1 |              | |                  |  V  Rotate all
+/// E  |       |              |<|                  ^  |
+/// F             | B Set |                        |  V
+/// G             |   f   |                        |<-|
+///
+/// Above is an illustration of how the various operations work for a f = 2 topology.
+
+#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Serialize, IntoSchema)]
 pub struct Topology {
     /// Current order of peers. The roles of peers are defined based on this order.
-    sorted_peers: Vec<PeerId>,
+    pub sorted_peers: Vec<PeerId>,
 }
 
 impl Topology {
-    /// Get Builder struct.
-    pub fn builder() -> Builder {
-        Builder::new()
-    }
-
-    /// Apply a `view change`, i.e. change the topology in case there were faults in the consensus round.
-    #[allow(clippy::expect_used)]
-    pub fn rebuild_with_new_view_change_count(&mut self, view_change_count: u64) {
-        *self = Builder::new().with_peers(self.sorted_peers.iter().cloned()).build(view_change_count).expect(
-            "Invalid topology. At this stage the error is unrecoverable. Please restart the peer.",
-        )
+    /// Create a new topology.
+    pub fn new<T: IntoIterator<Item = PeerId>>(peers: T) -> Self {
```

Comment:
```suggestion
    pub fn new(peers: impl IntoIterator<Item = PeerId>) -> Self {
```

Minor point, but this is the preferred approach

Reply:
Although, sometimes, it's better to do `impl FromIterator for ...`